dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/CompanyRolesControllerTest.kt

```diff
@@ -282,4 +288,123 @@ class CompanyRolesControllerTest {
+                hasUserCompanyRole(it, companyId, dataUploaderUserId)
+            }
+            assertErrorCodeInCommunityManagerClientException(exceptionWhenCheckingIfUserIsCompanyOwner, 404)
+        }
+    }
+
+    @Test
+    fun `assure that a company member admin without admin rights can only add and remove members and member admins`() {
+        val companyId = uploadCompanyAndReturnCompanyId()
+        val listOfRolesThatCanBeModified = listOf(CompanyRole.MemberAdmin, CompanyRole.Member)
+        val listOfRolesThatCannotBeModified =
+            listOf(CompanyRole.CompanyOwner, CompanyRole.DataUploader)
+        jwtHelper.authenticateApiCallsWithJwtForTechnicalUser(TechnicalUser.Admin)
+        assignCompanyRole(CompanyRole.MemberAdmin, companyId, dataReaderUserId)
+        listOfRolesThatCanBeModified.forEach {
+            jwtHelper.authenticateApiCallsWithJwtForTechnicalUser(TechnicalUser.Reader)
+            assignCompanyRole(it, companyId, dataUploaderUserId)
+
+            jwtHelper.authenticateApiCallsWithJwtForTechnicalUser(TechnicalUser.Uploader)
+            assertDoesNotThrow { hasUserCompanyRole(it, companyId, dataUploaderUserId) }
+
+            jwtHelper.authenticateApiCallsWithJwtForTechnicalUser(TechnicalUser.Reader)
+            removeCompanyRole(it, companyId, dataUploaderUserId)
+
+            val exceptionWhenCheckingIfUserIsCompanyOwner = assertThrows<ClientException> {
+                hasUserCompanyRole(it, companyId, dataUploaderUserId)
+            }
+            assertErrorCodeInCommunityManagerClientException(exceptionWhenCheckingIfUserIsCompanyOwner, 404)
+        }
+
+        listOfRolesThatCannotBeModified.forEach {
+            jwtHelper.authenticateApiCallsWithJwtForTechnicalUser(TechnicalUser.Reader)
+            val exceptionWhenTryingToAddCompanyMembers = assertThrows<ClientException> {
+                assignCompanyRole(it, companyId, dataUploaderUserId)
+            }
+            assertErrorCodeInCommunityManagerClientException(exceptionWhenTryingToAddCompanyMembers, 403)
+        }
+    }
+
+    @Test
+    fun `assure that a user that have member or uploader role or no role at all cannot add or remove company roles`() {
+        val companyId = uploadCompanyAndReturnCompanyId()
+        jwtHelper.authenticateApiCallsWithJwtForTechnicalUser(TechnicalUser.Reader)
+        tryToAssignAndRemoveCompanyMembersAndAssertThatItsForbidden(companyId)
+
+        val listOfCompanyRolesWithoutModificationRights =
+            listOf(CompanyRole.DataUploader, CompanyRole.Member)
+
+        listOfCompanyRolesWithoutModificationRights.forEach {
+            jwtHelper.authenticateApiCallsWithJwtForTechnicalUser(TechnicalUser.Admin)
+            assignCompanyRole(it, companyId, dataReaderUserId)
+            tryToAssignAndRemoveCompanyMembersAndAssertThatItsForbidden(companyId)
+        }
+    }
+
+    @Test
+    fun `assure that every company role has access to get and head endpoint but not a keycloak reader`() {
```

Comment:
Hier würde ich noch in den Text irgendwie die Info reinbringen, dass es gerade darum geht, dass die user mit den company roles nur auf ihre EIGENE company zugreifen können über GET und HEAD

oder?

Weil für eine companyId, wofür sie keine Rolle haben, hätten sie nach wie vor keinen Zugriff.

Vielleicht kannst du dieses Detail noch in die Test-Beschreibung kriegen irgendwie

Reply:
Done

Reply:
Ich finde die Info ist immer noch nicht in der Testbeschreibung.

Dort steht:
`assure that every company role has access to get and head endpoint but not a user without any company role`

besser wäre z.B.

`assure that every company role has access to get and head endpoint for the respective company even without admin keycloak role`

Also es soll direkt in der Beschreibung stehen, dass der springede Punkt ist, dass man für die EIGENE company (d.h. für die company, für die deine company role assignment gilt) an den GET und HEAD kommt, obwhol man kein keycloak admin ist.
Weil grundsätzlich können ja nur keycloak-admins an den GET und HEAD Endpunkt

Reply:
Ist das der Punkt? Ich dachte es geht um die Differenzierung zwischen data_reader Usern und data_reader Usern, die aber zu der Company gehören. Und das der Test den Unterschied zwischen diesen beiden Fällen testen und sicherstellen soll.